### PR TITLE
[WIP] UI: endpoint for init

### PR DESCRIPTION
**Pull request by rishabh3112**
**What kind of change does this PR introduce?**

feature
**Did you add tests for your changes?**
No
**If relevant, did you update the documentation?**
No
**Summary**
Started new PR instead for #685 as it was suppose to solve the cause using child_process, This one focus to solve programatically.

_For Contributors_
I have started writing up the init endpoint by utilising the init-generator present in `@webpack-cli/generators` package.

Things that are needed besides this:
1. ~Get user's root folder location to the server ( w/o explicitly asking user while questioning).~
*Done at 8e5cceb* 🎉

2. Client side intermediate mapper which maps question to its component in preact.
*Example: when questioner send `{action: "question", question: Input(....)}` should map to a component with template to a input field*
By @renatoagds

Linking parent issue #760 #649.

**Does this PR introduce a breaking change?**

**Other information**
cc: @renatoagds

### CODE REVIEW

File: packages/ui-server/utils/user.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
+    return new (P || (P = Promise))(function (resolve, reject) {
+        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
+        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
+        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
+        step((generator = generator.apply(thisArg, _arguments || [])).next());
+    });
+};
+Object.defineProperty(exports, "__esModule", { value: true });
+const net_1 = require("net");
```
Comment (renatoagds) [inline packages/ui-server/utils/user.js]: @rishabh3112 why we're using a transpiled code? ins't easier to install this by npm?

File: packages/ui-server/utils/questioner.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
+"use strict";
+Object.defineProperty(exports, "__esModule", { value: true });
```
Comment (renatoagds) [inline packages/ui-server/utils/questioner.js]: @rishabh3112 why we need to define this as a `esModule`?

File: packages/ui-server/routes/api/_entry.js [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+	} else {
+		return questioner.question({
+			action: "question",
+			question: "Which module will be the first to enter the application?"
+		}).then((singularEntryAnswer) => {
+			let singularEntry = singularEntryAnswer;
+			if (singularEntry.indexOf("\"") >= 0) {
+				singularEntry = singularEntry.replace(/"/g, "'");
+			}
+			return singularEntry;
```
Comment (renatoagds) [inline packages/ui-server/routes/api/_entry.js]: @rishabh3112 we could reduce this to:

```
return singularEntryAnswer.indexOf("\"") >= 0
  ? singularEntryAnswer.replace(/"/g, "'")
  : singularEntryAnswer;
```

File: packages/ui-server/routes/api/_entry.js [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+				});
+			}).then((entryPropAnswer) => {
+				Object.keys(entryPropAnswer).forEach(val => {
+					if (
+						entryPropAnswer[val].charAt(0) !== "(" &&
+						entryPropAnswer[val].charAt(0) !== "[" &&
+						!entryPropAnswer[val].includes("function") &&
+						!entryPropAnswer[val].includes("path") &&
+						!entryPropAnswer[val].includes("process")
+					) {
```
Comment (renatoagds) [inline packages/ui-server/routes/api/_entry.js]: @rishabh3112 we have this verification in two places, here and in forEachPromise, you could extract it to a helper.

Comment (renatoagds) [reply packages/ui-server/routes/api/_entry.js]: btw, which cases we're trying to avoid with this verification?

File: packages/ui-server/routes/api/_entry.js [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+				});
+			} else {
+				n = {};
+			}
+			return fn(trimmedProp);
+		});
+	}, Promise.resolve());
+}
+
+module.exports = function(questioner, answer) {
```
Comment (renatoagds) [inline packages/ui-server/routes/api/_entry.js]: @rishabh3112 looking first in the code, I didn't know what `answer` mean. After, reading the code I understood that is an answer for the question about multiple bundle. We could rename this for `usingMultipleBundles`, it will reduce the time to understand this part.

File: packages/ui-server/routes/api/init.js [hunk 6]
```diff
@@ -33,7 +33,19 @@ router.get("/", function(req, res, next) {
...
+		return entryQuestions(questioner, answer);
+	}).then((entryOptions) => {
+		if (entryOptions !== "\"\"") {
+			config.setWebpackOption("webpackOptions", {
+				entry: entryOptions,
+			});
+		}
+		return questioner.question({
+			action: "question",
+			question: "Which folder will Which folder will your generated bundles be in? [default: dist]"
```
Comment (renatoagds) [inline packages/ui-server/routes/api/init.js]: duplication in the phrase here.

File: packages/ui-server/utils/user.js (hunk 1 above)
Comment (rishabh3112) [reply packages/ui-server/utils/user.js]: Commited by mistake

File: packages/ui-server/utils/questioner.js (hunk 2 above)
Comment (rishabh3112) [reply packages/ui-server/utils/questioner.js]: Same here

File: packages/ui-server/routes/api/init.js (hunk 6 above)
Comment (rishabh3112) [reply packages/ui-server/routes/api/init.js]: I don't get what you want to say here

Comment (renatoagds) [reply packages/ui-server/routes/api/init.js]: **Wich folder will Which folder will** ... The begin of the phrase is duplicated

File: packages/ui-server/routes/api/_entry.js (hunk 4 above)
Comment (rishabh3112) [reply packages/ui-server/routes/api/_entry.js]: These are from the `init-generator` from generator package.
BTW,
These are here to avoid user to give functional input instead of static input

File: packages/ui-server/routes/api/init.js (hunk 6 above)
Comment (rishabh3112) [reply packages/ui-server/routes/api/init.js]: Okay I got it now 😂

File: packages/ui-server/routes/api/_entry.js (hunk 5 above)
Comment (rishabh3112) [reply packages/ui-server/routes/api/_entry.js]: Sure will make it once the port is complete.
Thanks for pointing out 😀

File: packages/ui-server/routes/api/_entry.js (hunk 3 above)
Comment (rishabh3112) [reply packages/ui-server/routes/api/_entry.js]: Yeah this can be done, but I want to maintain as much similarity as possible with the prompting of init-generator so it is easier to manipulate here when munipulated at init-generator.

Comment (renatoagds) [reply packages/ui-server/routes/api/_entry.js]: @rishabh3112 why do not change the `init-generator` too?

Comment (rishabh3112) [reply packages/ui-server/routes/api/_entry.js]: Not there is use case on checking input length else use the default value. This case isn't valid here as this must be handled on client side for UI.

Comment (renatoagds) [reply packages/ui-server/routes/api/_entry.js]: @rishabh3112 sorry I didn't get your point. Here is just a readability and code style change, it won't change nothing in the way of the code works.

Comment (rishabh3112) [reply packages/ui-server/routes/api/_entry.js]: Have a look at init-generator, you would get what I want to say here 👍

**Comment by rishabh3112** [comment]
Hey @renatoagds are you willing to work on anty of the issues mentioned in PR description summary ?

**Comment by renatoagds** [comment]
@rishabh3112 I will work on 2 [connection with client side]. 👍

**Comment by rishabh3112** [comment]
I am incorporating inquirer questions in next commit so design mapper as per inquirer questions🙂

File: packages/ui-server/routes/api/_entry.js (hunk 5 above)
Comment (rishabh3112) [reply packages/ui-server/routes/api/_entry.js]: JsDoc is added for this for now

File: packages/ui-server/routes/api/init.js [hunk 7]
```diff
@@ -31,10 +38,337 @@ router.get("/", function(req, res, next) {
...
+	let regExpForStyles;
+
 	questioner.start({
 		action: "question",
-		question: "A dummy question for reference"
-	}).then((data) => {
+		question: Confirm("entryType", "Will your application have multiple bundles?"),
+	}).then((entryTypeAnswer) => {
+		return entryQuestions(questioner, entryTypeAnswer);
+	}).then((entryOptions) => {
```
Comment (ematipico) [inline packages/ui-server/routes/api/init.js]: I guess this block is taken form the `init` package, correct?

Comment (rishabh3112) [reply packages/ui-server/routes/api/init.js]: Yeah, whole endpoint is just JS port of init-generator @ematipico with instead of using yeoman prompts, TCP classes are used which I designed to have almost the same API.

File: packages/ui/bin/www.js [hunk 8]
```diff
@@ -4,7 +4,10 @@
  */
 
 const path = require("path");
-const app = require("../../ui-server")(path.join(__dirname, "../../ui-gui/build")); // eslint-disable-line
+const STATIC = path.existsSync(path.join(__dirname, "../../ui-gui/build"))
```
Comment (renatoagds) [inline packages/ui/bin/www.js]: @rishabh3112 `path.existsSync` doesn't exists, you need to use `fs.existsSync`

**Comment by renatoagds** [comment]
@rishabh3112 could please explain how can I test this init entry? I would like to start think about the UI connection with it. btw, @evenstensberg any update about zalando designers?

File: packages/ui/bin/www.js (hunk 8 above)
Comment (rishabh3112) [reply packages/ui/bin/www.js]: Ohh my bad.

**Comment by evenstensberg** [comment]
@renatoagds zalando designers didn't have time, but if you can find the design issue we're tracking progress there

**Comment by sohailrajdev97** [comment]
@rishabh3112, I fetched this PR locally for testing. After cloning the repository, installing dependencies, bootstraping the packages and building, everything was functional. After fetching this PR in a local branch, I followed these steps:

1. Built the `ui-gui` package
1. Did an `npm install`

The following error was encountered during `npm install`:

```
sohailrajdev@:~/Developer/webpack-cli/: npm install

> webpack-cli@3.1.2 postinstall /Users/sohailrajdev/Developer/webpack-cli
> lerna bootstrap

lerna notice cli v3.4.0
lerna info Bootstrapping 16 packages
lerna info Installing external dependencies
lerna ERR! npm install exited 1 in '@webpack-cli/ui'
lerna ERR! npm install stderr:
npm ERR! code E404
npm ERR! 404 Not Found: @webpack-cli/ui-gui@0.0.1
```
Can someone please help me with this issue ?

**Comment by rishabh3112** [comment]
UI Packages are not published to npm that's why this error.

**Comment by sohailrajdev97** [comment]
@

> UI Packages are not published to npm that's why this error.

Thanks for a quick reply @rishabh3112. Can you please tell me a way for locally testing this PR. When I do `node packages/ui/bin/www.js ` and try sending a request to `localhost:3000/api/init`, The following error is reported by the express server:

```
Error: Cannot find module '@webpack-cli/generators/init-generator'
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:594:15)
    at Function.Module._load (internal/modules/cjs/loader.js:520:25)
    at Module.require (internal/modules/cjs/loader.js:650:17)
    at require (internal/modules/cjs/helpers.js:20:18)
    at Object.<anonymous> (/Users/sohailrajdev/Developer/webpack-cli/packages/init/index.js:3:26)
    at Module._compile (internal/modules/cjs/loader.js:702:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:713:10)
    at Module.load (internal/modules/cjs/loader.js:612:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:551:12)
    at Function.Module._load (internal/modules/cjs/loader.js:543:3)
```

**Comment by rishabh3112** [comment]
Because you have not linked `init` package.

**Comment by fokusferit** [comment]
Hey @renatoagds today during our cli meeting @rishabh3112 mentioned that you're working on the mapper. Do you have any update or need some help, clarity? 

Would be great if we can try to merge this PR soon.

**Comment by renatoagds** [comment]
@fokusferit Hey. I asked a question for @rishabh3112 here: https://github.com/webpack/webpack-cli/pull/756#issuecomment-467241450 ... I ran the server, and start the mappers but I need to simulate the calls to init endpoint, I need some help to do that, with that I could go ahead and finish the mapper in preact.

**Comment by rishabh3112** [comment]
Hi @renatoagds Use object of Client class from utils for this and send parameters as you recieve from endpoint in GET call.

**Comment by renatoagds** [comment]
@rishabh3112 Thanks! I'll try it today at night and will try to return tomorrow with a initial mapper.

**Comment by rishabh3112** [comment]
User directory issue is solved and now the UI is working like a charm 🎉. Seeig conf generated where UI ran:
```js
const webpack = require('webpack')
const path = require('path')




/*
* SplitChunksPlugin is enabled by default and replaced
* deprecated CommonsChunkPlugin. It automatically identifies modules which
* should be splitted of chunk by heuristics using module duplication count and
* module category (i. e. node_modules). And splits the chunks…
*
* It is safe to remove "splitChunks" from the generated configuration
* and was added as an educational example.
*
* https://webpack.js.org/plugins/split-chunks-plugin/
*
*/

/*
* We've enabled UglifyJSPlugin for you! This minifies your app
* in order to load faster and run less javascript.
*
* https://github.com/webpack-contrib/uglifyjs-webpack-plugin
*
*/

const UglifyJSPlugin = require('uglifyjs-webpack-plugin');




module.exports = {
  module: {
    rules: [{
      include: [path.resolve(__dirname, 'src')],
      loader: 'babel-loader',

      options: {
        plugins: ['syntax-dynamic-import'],

        presets: [['env', {
          'modules': false
        }]]
      },

      test: /\.js$/
    }, {
      test: /\.(scss|css)$/,

      use: [{
        loader: 'style-loader'
      }, {
        loader: 'css-loader'
      }, {
        loader: 'sass-loader'
      }]
    }]
  },

  entry: "./src/index",

  output: {
    filename: '[name].[chunkhash].js',
    path: path.resolve(__dirname, 'dist')
  },

  mode: 'development',

  optimization: {
    splitChunks: {
      cacheGroups: {
        vendors: {
          priority: -10,
          test: /[\\/]node_modules[\\/]/
        }
      },

      chunks: 'async',
      minChunks: 1,
      minSize: 30000,
      name: true
    }
  },

  merge: {}
}
```
We just need a mapper now that translate Inquirer questions to components for first prototype :)

Just wanted to know one thing @evenstensberg by `merge` object is added under an property and not merged with the final config itself (like above) ? Or webpack takes care of it while building ?

File: packages/ui-server/routes/api/_tooltip.js [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+* To enable SASS or LESS, add the respective loaders to module.rules
+*
+* https://github.com/postcss/postcss
+*
+* https://github.com/postcss/autoprefixer
+*
+* https://github.com/jonathantneal/precss
+*
+*/`,
+	uglify: () => `/*
```
Comment (evenstensberg) [inline packages/ui-server/routes/api/_tooltip.js]: We're using terser

**Comment by rishabh3112** [comment]
>  We need to reuse generator logic so we avoid a lot of spagetti code. 

You want to generator classes to serve the endpoint? That's a lot of work to do from this point. We need to wrap generator classes with the TCP classes that I made for prompting. I can try :)

> UI endpoints should just be base paths for the dashboard, etc paths, nothing more.

The routes for the dashboard, etc. would be made using client-side routing and that makes more sense to me. Having an api endpoint for each route is insane and will make server unmaintainable soon. The server should serve REST API for any data required by the front side like questionaire, stats, etc.
In short:
`ui-server`: API with data to populate GUI and run node processes
`ui-gui`: Accepts data and use API to invoke specific node process

Consider those points and I will do as you say 👍

**Comment by rishabh3112** [comment]
On the other way around we can also accept just a function instead of class for scaffolding using CLI by mapping that function to prompting in a sample Yeoman class. That same function can also be used at UI to serve questionaire.

File: packages/ui-server/routes/api/_tooltip.js (hunk 9 above)
Comment (rishabh3112) [reply packages/ui-server/routes/api/_tooltip.js]: Yeah, will do it here when terser PR get merged in master 👍

**Comment by rishabh3112** [comment]
Should I follow my work on the current approach then @evenstensberg ?

**Comment by evenstensberg** [comment]
Yes

**Comment by rishabh3112** [comment]
Closing PR in favour of new strategy!!
